Add missing ascend types

Fabi2607 opened the pull request:
The tour book also allows additional ascend styles for looking at and wanting to climb routes.

paetz commented on app/src/main/java/com/example/paetz/yacguide/database/Ascend.java:
```diff
@@ -27,6 +27,10 @@
         styles.put(7, "Nachstieg");
         styles.put(8, "Hinterhergehampelt");
         styles.put(9, "Gesackt");
+        styles.put(10, "zugesehen");
```
Hi, sorry für die späte Antwort.

Ich habe diese Stile absichtlich weggelassen, da
1) die Stilauswahl beim Eintragen einer Begehung dann noch unübersichtlicher wird als sie jetzt schon ist,
2) die Grünfärbung des Weges nach eingetragener Begehung ihren Sinn verliert. Sie bedeutet "Weg wurde abgehakt (zumindest teilweise)". Man könnte über verschiedene Farben nachdenken, zB gelb für "markiert" = "will ich klettern" oder so.

Gibt es einen Grund, wieso du die Sachen gern rein haben willst?

Fabi2607 replied:
Ich habe mehrere Gründe:
1. Die Anzeige dieser Begehungsarten funktioniert nicht, wenn man sie von der Webseite exportiert
2. Ich benutze die DB gerne um mir schonmal Wege zu markieren, die ich bald klettern will, gerne auch unterwegs mittels App

Ich habe bereits eine neue Version, die Anstatt der Einfärbung Unicode Symbole verwendet
(Häkchen/Kreuz für geklettert/gesackt, Stift für Merkliste), da ich die gleichen Gedanken zur Einfärbung hatte.

paetz commented:
Ja ich denke, das ist sinnvoll und wäre top, wenn du das (evtl. als neuen PR) pushen könntest.
Folgender Wunsch meinerseits:
1. Ich fänd es besser, wenn das Tourenbuch nur gekletterte Wege beinhaltet (Häkchen).
2. Projekte (="will ich klettern") und Säcke sollten jeweils in separaten Listen sein, die man vom Hauptmenü aus erreicht. Die Symbole kannst du in diesen Listen dann weglassen.
3 Die übrigen Stile ("Zugesehen", "An den Einstieg gepinkelt", "von gehört") würde ich nicht in einer extra Liste halten und in der Wegliste mit einem Augen-Symbol symbolisieren.
4. Falls du ein neues .apk baust, bitte einen neuen Versions-Ordner im release-Order anzulegen und API-Version 22 (=Android 5.1) nutzen

paetz commented on app/src/main/java/com/example/paetz/yacguide/utils/WidgetUtils.java:
```diff
@@ -78,4 +78,16 @@ private static TextView _createTextView(Context context, String text, int textSi
+            symbol = "✓";
+        } else if (styleId == 9) {
+            symbol = "✗";
+        } else if (styleId < 14){
```
Ascend styles "an den einstieg gepinkelt", "zugesehen" and "von gehört" should have an eye-symbol (1F441).

Fabi2607 replied:
> Ja ich denke, das ist sinnvoll. Folgendes fänd ich noch gut:
> 
>     1. Ich fänd es besser, wenn das Tourenbuch nur gekletterte Wege beinhaltet (Häkchen).
> 
>     2. Projekte (="will ich klettern") und Säcke sollten jeweils in separaten Listen sein, die man vom Hauptmenü aus erreicht. Die Symbole kannst du in diesen Listen dann weglassen.
> 
> 
> Das sollte man aber vllt als separaten PR machen. Willst du das auch noch machen oder soll ich?

Ich gucke es mir mal an und versuche mich mal dran.

paetz commented on app/src/main/java/com/example/paetz/yacguide/TourbookActivity.java:
```diff
@@ -161,8 +161,9 @@ public void onClick(View v) {
+
             layout.addView(WidgetUtils.createCommonRowLayout(this,
-                    rock.getName() + " - " + route.getName(),
+                    WidgetUtils.getAscendSymbol(ascend.getStyleId()) + " " + rock.getName() + " - " + route.getName(),
```
Having done an ascend and having a project at the same time on the same rock should be possible. --> Multiple symbols would be needed here.

Fabi2607 replied:
OK, I will allow each symbol to show up exactly once per summit.

paetz replied:
Sorry, my mistake. The tourbook entry should not be affected by this.

paetz commented on app/src/main/java/com/example/paetz/yacguide/RouteActivity.java:
```diff
@@ -115,12 +118,20 @@ public void onClick(View v) {
+
             layout.addView(WidgetUtils.createCommonRowLayout(this,
-                    route.getName() + commentCountAddon,
+                    WidgetUtils.getAscendSymbol(lowestAscendId) + " " + route.getName() + commentCountAddon,
```
Imagine you have a follow-ascend and a lead-"Sack" at the same route. --> Multiple symbols must be possible.

paetz replied:
The information about ascended / sacked / watched / projected should be stored in the db field in Route table (can we hack field "ascendCount" for that?). Doing so, we prevent iterating over all ascends each time we enter a route list.

Fabi2607 replied:
I thought about that, but my knowledge about the framework is very limited and I did not want to break the database compatibility.

paetz commented on app/src/main/java/com/example/paetz/yacguide/utils/WidgetUtils.java:
```diff
@@ -78,4 +78,16 @@ private static TextView _createTextView(Context context, String text, int textSi
+
+    public static String getAscendSymbol(int styleId) {
+        String symbol = " ";
+        if (styleId < 9) {
```
We should not operate on ids since this is error-prone and not easy to follow. Please create an enum.

Fabi2607 replied:
Sure, makes sense.

paetz replied:
Best would be a separate class AscendStyles:
```suggestion
      public class AscendStyles {
           private class AscendStyle { 
                 int id;
                 String name;
                 String symbol;
             }
              // init static ascendStyle list here
             
              public List<String> getAscendNames() {
                   // return all names as string list
              } 
              public int getAscendId(String name) { ... }
              
              public String getAscendSymbol(int id) { ... }
         }
```

Fabi2607 replied:
That approach would also work. I went for a simpler variant, but can change it, if you prefer.

paetz commented on app/src/main/java/com/example/paetz/yacguide/database/Ascend.java:
```diff
@@ -27,6 +27,10 @@
+        styles.put(10, "zugesehen");
+        styles.put(11, "an den Einstieg gepinkelt");
+        styles.put(12, "von gehört");
+        styles.put(13, "Will ich klettern");
```
The other new ascend styles should also start with a capital letter.

paetz commented on app/src/main/java/com/example/paetz/yacguide/utils/WidgetUtils.java:
```diff
@@ -78,4 +78,16 @@ private static TextView _createTextView(Context context, String text, int textSi
+        } else if (styleId == 9) {
+            symbol = "✗";
+        } else if (styleId < 14){
+            symbol = "✎";
```
For "project", I would prefer this one: U+261D

Fabi2607 commented:
- I grouped all successful/unsuccessful attempts in the tourbook. (If you bailed you atleast tried to climb the route).
- All variations of marking the route are now in a separate ticklist.
- I addressed all the comments, except for the ones I only read after I pushed the newest revision (DB related changes)

paetz commented on app/src/main/java/com/example/paetz/yacguide/MainActivity.java:
```diff
@@ -29,4 +30,10 @@ public void enterTourbook(View v) {
+
+    public void enterTicklist(View v) {
+        Intent intent = new Intent(this, TourbookActivity.class);
+        intent.putExtra(IntentConstants.IS_TICKLIST_KEY, true);
```
We should pass an enum or int type for tourbook-, säcke- and project-distinction. --> Done in follow-up ticket #52.

paetz commented on app/src/main/java/com/example/paetz/yacguide/TourbookActivity.java:
```diff
@@ -113,13 +118,18 @@ public void export(View v) {
         int currentMonth, currentDay, currentRegionId;
         currentMonth = currentDay = currentRegionId = -1;
         for (final Ascend ascend : ascends) {
+            if (_isTicklist && ascend.getStyleId() <= AscendTypes.BAILED || !_isTicklist && ascend.getStyleId() > AscendTypes.BAILED) {
```
Should be encapsulated in a method. --> Done in follow-up ticket #52.

paetz commented on app/src/main/java/com/example/paetz/yacguide/TourbookActivity.java:
```diff
@@ -161,8 +171,9 @@ public void onClick(View v) {
+
             layout.addView(WidgetUtils.createCommonRowLayout(this,
-                    rock.getName() + " - " + route.getName(),
+                    WidgetUtils.getAscendSymbol(ascend.getStyleId()) + " " + rock.getName() + " - " + route.getName(),
```
No symbol needed when we have a list for Säcke. --> Done in follow-up ticket #52.

paetz commented on app/src/main/java/com/example/paetz/yacguide/network/RockParser.java:
```diff
@@ -109,7 +111,9 @@ private void parseRoutes(String json) throws JSONException {
             for (final Route route : db.routeDao().getAll(rock.getId())) {
-                ascended |= (route.getAscendCount() > 0);
+                for (final Ascend ascend : db.ascendDao().getAscendsForRoute(route.getId())) {
+                    ascended |= (ascend.getStyleId() < AscendTypes.BAILED);
```
Info about ascend type could be stored in Route db field ascendCount as binary digits to prevent iteration over (probably lots of) ascends, e.g. 0b1001 = ascended (1), not sacked (0), not watched (0), projected (1). Same for RouteActivity. --> Done in follow-up ticket #52.

paetz commented on app/src/main/java/com/example/paetz/yacguide/utils/WidgetUtils.java:
```diff
@@ -78,4 +83,33 @@ private static TextView _createTextView(Context context, String text, int textSi
+    public static String getAscendSymbol(int styleId) {
+        String symbol = " ";
+
+        if (styleId < AscendTypes.BAILED) {
```
To prevent this cost-intense if-else-branching for every route's ascend, symbol should be stored in a dedicated class AscendType together with id and name. --> Done in follow-up ticket #52.

paetz commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -35,6 +35,15 @@
+        android:layout_below="@id/tourbookButton"
+        android:layout_centerInParent="true"
+        android:onClick="enterTicklist"
+        android:text="Merkliste" />
```
Should be named "Projekte". --> Done in follow-up ticket #52.
